File: sdk/npm/src/providers/zhujump.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";
import { randomInt } from "crypto";

const BASE_URL = "https://mail.zhujump.com";
const LOGIN_REFERER = `${BASE_URL}/zh-CN/login`;
const TOKEN_PREFIX = "zhj1:";
const DEFAULT_EXPIRY_TIME = 60 * 60 * 1000;

const JSON_HEADERS: Record<string, string> = {
  Accept: "application/json",
  Origin: BASE_URL,
  Referer: LOGIN_REFERER,
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36",
};

interface SessionToken {
  c: string;
  i: string;
  b?: string;
}

interface MoemailInstance {
  baseUrl: string;
  domain: string;
  channel: Channel;
  expiryTime?: number | null;
}

function randomValue(prefix: string, size: number): string {
  const chars = "abcdefghijklmnopqrstuvwxyz0123456789";
  let out = prefix;
  for (let i = 0; i < size; i++) {
    out += chars[randomInt(chars.length)];
  }
  return out;
}

function randomPassword(): string {
  return `Sdk!${randomValue("", 12)}A1`;
}

function setCookieLines(headers: Headers): string[] {
  const h = headers as Headers & { getSetCookie?: () => string[] };
  if (typeof h.getSetCookie === "function") {
    return h.getSetCookie();
  }
  const one = headers.get("set-cookie");
  return one ? one.split(/,(?=[^ ;]+=)/) : [];
}

function mergeCookieHeader(prev: string, headers: Headers): string {
  const jar = new Map<string, string>();
  for (const part of prev.split(";")) {
    const text = part.trim();
    const idx = text.indexOf("=");
    if (idx > 0) {
      jar.set(text.slice(0, idx), text.slice(idx + 1));
    }
  }
  for (const line of setCookieLines(headers)) {
    const first = line.split(";")[0]?.trim() || "";
    const idx = first.indexOf("=");
    if (idx > 0) {
      jar.set(first.slice(0, idx), first.slice(idx + 1));
    }
  }
  return [...jar.entries()].map(([k, v]) => `${k}=${v}`).join("; ");
}

function encodeToken(token: SessionToken): string {
  return (
    TOKEN_PREFIX + Buffer.from(JSON.stringify(token), "utf8").toString("base64")
  );
}

function decodeToken(token: string): SessionToken {
  if (!token.startsWith(TOKEN_PREFIX)) {
    throw new Error("zhujump: invalid session token");
  }

  let raw = "";
  try {
    raw = Buffer.from(token.slice(TOKEN_PREFIX.length), "base64").toString(
      "utf8",
    );
  } catch {
    throw new Error("zhujump: invalid session token");
  }

  const parsed = JSON.parse(raw) as Partial<SessionToken>;
  const cookie = String(parsed.c || "").trim();
  const emailId = String(parsed.i || "").trim();
  if (!cookie || !emailId) {
    throw new Error("zhujump: invalid session token");
  }
  const baseUrl = String(parsed.b || "").trim();
  return baseUrl
    ? { c: cookie, i: emailId, b: baseUrl }
    : { c: cookie, i: emailId };
}

function loginReferer(baseUrl: string): string {
  return `${baseUrl}/zh-CN/login`;
}

function jsonHeaders(baseUrl: string): Record<string, string> {
  return {
    ...JSON_HEADERS,
    Origin: baseUrl,
    Referer: loginReferer(baseUrl),
  };
}

async function ensureLoggedIn(
  baseUrl: string,
  cookie: string,
  username: string,
): Promise<void> {
  const response = await fetchWithTimeout(`${baseUrl}/api/auth/session`, {
    headers: {
      ...jsonHeaders(baseUrl),
      Cookie: cookie,
    },
  });
  if (!response.ok) {
    throw new Error(`zhujump session: ${response.status}`);
  }

  const json = (await response.json()) as {
    user?: {
      username?: string;
    };
  };
  if (String(json.user?.username || "").trim() !== username) {
    throw new Error("zhujump: login verification failed");
  }
}

async function loginAndCreate(
  instance: MoemailInstance,
): Promise<{ email: string; emailId: string; cookie: string }> {
  const username = randomValue("sdk", 10);
  const password = randomPassword();
  let cookie = "";
  const baseUrl = instance.baseUrl.replace(/\/$/, "");
  const headers = jsonHeaders(baseUrl);

  const registerResp = await fetchWithTimeout(`${baseUrl}/api/auth/register`, {
    method: "POST",
    headers: {
      ...headers,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ username, password, turnstileToken: "" }),
  });
  if (!registerResp.ok) {
    throw new Error(`zhujump register: ${registerResp.status}`);
  }
  cookie = mergeCookieHeader(cookie, registerResp.headers);

  const csrfResp = await fetchWithTimeout(`${baseUrl}/api/auth/csrf`, {
    headers: {
      ...headers,
      Cookie: cookie,
    },
  });
  if (!csrfResp.ok) {
    throw new Error(`zhujump csrf: ${csrfResp.status}`);
  }
  cookie = mergeCookieHeader(cookie, csrfResp.headers);

  const csrfJson = (await csrfResp.json()) as { csrfToken?: string };
  const csrfToken = String(csrfJson.csrfToken || "").trim();
  if (!csrfToken) {
    throw new Error("zhujump: csrf token missing");
  }

  const loginResp = await fetchWithTimeout(
    `${baseUrl}/api/auth/callback/credentials?`,
    {
      method: "POST",
      headers: {
        ...headers,
        Cookie: cookie,
        "Content-Type": "application/x-www-form-urlencoded",
        "x-auth-return-redirect": "1",
      },
      body: new URLSearchParams({
        username,
        password,
        turnstileToken: "",
        redirect: "false",
        csrfToken,
        callbackUrl: loginReferer(baseUrl),
      }).toString(),
      redirect: "manual",
    },
  );
  if (!loginResp.ok) {
    throw new Error(`zhujump login: ${loginResp.status}`);
  }
  cookie = mergeCookieHeader(cookie, loginResp.headers);

  await ensureLoggedIn(baseUrl, cookie, username);

  const local = randomValue("sdk", 10);
  const body: Record<string, string | number> = {
    name: local,
    domain: instance.domain,
  };
  if (instance.expiryTime !== null) {
    body.expiryTime = instance.expiryTime ?? DEFAULT_EXPIRY_TIME;
  }

  const generateResp = await fetchWithTimeout(
    `${baseUrl}/api/emails/generate`,
    {
      method: "POST",
      headers: {
        ...headers,
        Cookie: cookie,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    },
  );
  if (!generateResp.ok) {
    throw new Error(`zhujump generate: ${generateResp.status}`);
  }

  const generateJson = (await generateResp.json()) as {
    id?: string;
    email?: string;
  };
  const emailId = String(generateJson.id || "").trim();
  const email = String(generateJson.email || "").trim();
  if (!emailId || !email) {
    throw new Error("zhujump: invalid generate response");
  }

  return { email, emailId, cookie };
}

/**
 * 创建 mail.zhujump.com 固定域临时邮箱
 */
export async function generateEmail(
  domain: string,
  channel: Channel,
): Promise<InternalEmailInfo> {
  return generateEmailForInstance({
    baseUrl: BASE_URL,
    domain,
    channel,
    expiryTime: null,
  });
}

/**
 * 创建指定 MoeMail 部署实例的临时邮箱
 */
export async function generateEmailForInstance(
  instance: MoemailInstance,
): Promise<InternalEmailInfo> {
  const baseUrl = instance.baseUrl.replace(/\/$/, "");
  const result = await loginAndCreate({ ...instance, baseUrl });
  return {
    channel: instance.channel,
    email: result.email,
    token: encodeToken({ c: result.cookie, i: result.emailId, b: baseUrl }),
  };
}

async function fetchMessageDetail(
  baseUrl: string,
  cookie: string,
  emailId: string,
  messageId: string,
): Promise<Record<string, unknown> | null> {
  const response = await fetchWithTimeout(
    `${baseUrl}/api/emails/${encodeURIComponent(emailId)}/${encodeURIComponent(messageId)}`,
    {
      headers: {
        ...jsonHeaders(baseUrl),
        Cookie: cookie,
      },
    },
  );
  if (!response.ok) return null;
  const json = (await response.json()) as { message?: Record<string, unknown> };
  return json.message && typeof json.message === "object" ? json.message : null;
}

function hasMessageBody(row: { content?: string; html?: string }): boolean {
  return Boolean(
    String(row.content || "").trim() || String(row.html || "").trim(),
  );
}

/**
 * 获取 mail.zhujump.com 邮件列表
 */
export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  const session = decodeToken(token);
  const baseUrl = (session.b || BASE_URL).replace(/\/$/, "");
  const response = await fetchWithTimeout(
    `${baseUrl}/api/emails/${encodeURIComponent(session.i)}`,
    {
      headers: {
        ...jsonHeaders(baseUrl),
        Cookie: session.c,
      },
    },
  );
  if (!response.ok) {
    throw new Error(`zhujump get emails: ${response.status}`);
  }

  const json = (await response.json()) as {
    messages?: Array<{
      id?: string;
      from_address?: string;
      to_address?: string | null;
      subject?: string;
      content?: string;
      html?: string;
      sent_at?: number;
      received_at?: number;
    }>;
  };

  const rows = Array.isArray(json.messages) ? json.messages : [];
  const emails: Email[] = [];
  for (const row of rows) {
    let source: Record<string, unknown> = row;
    const id = String(row.id || "").trim();
    if (id && !hasMessageBody(row)) {
      const detail = await fetchMessageDetail(
        baseUrl,
        session.c,
        session.i,
        id,
      );
      if (detail) {
        source = { ...row, ...detail };
      }
    }
    emails.push(
      normalizeEmail(
        {
          id: source.id || "",
          from_address: source.from_address || "",
          to_address: source.to_address || email,
          subject: source.subject || "",
          content: source.content || "",
          html: source.html || "",
          received_at: source.received_at ?? source.sent_at,
          isRead: false,
        },
        email,
      ),
    );
  }
  return emails;
}

File: sdk/rust/src/providers/zhujump.rs
/*!
 * mail.zhujump.com 固定域渠道
 * 通过注册账号、登录会话后创建固定域邮箱，再通过邮箱 ID 获取邮件列表。
 */

use base64::Engine;
use rand::Rng;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use wreq::header::HeaderMap;

use crate::config::{block_on, get_current_ua, http_client_no_cookie_jar};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};

const BASE_URL: &str = "https://mail.zhujump.com";
const TOKEN_PREFIX: &str = "zhj1:";

#[derive(Serialize, Deserialize)]
struct ZhujumpToken {
    c: String,
    i: String,
    b: Option<String>,
}

fn random_value(prefix: &str, size: usize) -> String {
    const CHARS: &[u8] = b"abcdefghijklmnopqrstuvwxyz0123456789";
    let mut rng = rand::thread_rng();
    let mut out = String::from(prefix);
    for _ in 0..size {
        out.push(CHARS[rng.gen_range(0..CHARS.len())] as char);
    }
    out
}

fn random_password() -> String {
    format!("Sdk!{}A1", random_value("", 12))
}

fn login_referer(base_url: &str) -> String {
    format!("{}/zh-CN/login", base_url.trim_end_matches('/'))
}

fn json_headers(
    builder: wreq::RequestBuilder,
    base_url: &str,
    cookie: Option<&str>,
) -> wreq::RequestBuilder {
    let builder = builder
        .header("Accept", "application/json")
        .header("Origin", base_url.trim_end_matches('/'))
        .header("Referer", login_referer(base_url))
        .header("User-Agent", get_current_ua());
    if let Some(cookie) = cookie {
        if !cookie.trim().is_empty() {
            return builder.header("Cookie", cookie);
        }
    }
    builder
}

fn parse_cookie_header(hdr: &str) -> std::collections::BTreeMap<String, String> {
    let mut m = std::collections::BTreeMap::new();
    for part in hdr.split(';') {
        let part = part.trim();
        if part.is_empty() {
            continue;
        }
        if let Some(i) = part.find('=') {
            let k = part[..i].trim();
            let v = part[i + 1..].trim();
            if !k.is_empty() {
                m.insert(k.to_string(), v.to_string());
            }
        }
    }
    m
}

fn merge_set_cookies(hdr: &str, headers: &HeaderMap) -> String {
    let mut m = parse_cookie_header(hdr);
    for val in headers.get_all("set-cookie") {
        let Ok(s) = val.to_str() else {
            continue;
        };
        let first = s.split(';').next().unwrap_or("").trim();
        if let Some(i) = first.find('=') {
            let k = first[..i].trim();
            let v = first[i + 1..].trim();
            if !k.is_empty() {
                m.insert(k.to_string(), v.to_string());
            }
        }
    }
    m.into_iter()
        .map(|(k, v)| format!("{}={}", k, v))
        .collect::<Vec<_>>()
        .join("; ")
}

fn encode_token(cookie: &str, email_id: &str, base_url: &str) -> Result<String, String> {
    let raw = serde_json::to_vec(&ZhujumpToken {
        c: cookie.to_string(),
        i: email_id.to_string(),
        b: Some(base_url.trim_end_matches('/').to_string()),
    })
    .map_err(|e| format!("zhujump: token encode: {}", e))?;
    Ok(format!(
        "{}{}",
        TOKEN_PREFIX,
        base64::engine::general_purpose::STANDARD.encode(raw)
    ))
}

fn decode_token(token: &str) -> Result<ZhujumpToken, String> {
    if !token.starts_with(TOKEN_PREFIX) {
        return Err("zhujump: invalid session token".into());
    }
    let raw = base64::engine::general_purpose::STANDARD
        .decode(&token[TOKEN_PREFIX.len()..])
        .map_err(|_| "zhujump: invalid session token".to_string())?;
    let mut parsed: ZhujumpToken =
        serde_json::from_slice(&raw).map_err(|_| "zhujump: invalid session token".to_string())?;
    if parsed.c.trim().is_empty() || parsed.i.trim().is_empty() {
        return Err("zhujump: invalid session token".into());
    }
    if parsed.b.as_deref().unwrap_or("").trim().is_empty() {
        parsed.b = Some(BASE_URL.to_string());
    }
    Ok(parsed)
}

pub fn generate_email(domain: &str, channel: Channel) -> Result<EmailInfo, String> {
    generate_email_for_instance(BASE_URL, domain, channel, None)
}

pub fn generate_email_for_instance(
    base_url: &str,
    domain: &str,
    channel: Channel,
    expiry_time: Option<i64>,
) -> Result<EmailInfo, String> {
    let username = random_value("sdk", 10);
    let password = random_password();
    let local = random_value("sdk", 10);
    let base_url = base_url.trim_end_matches('/').to_string();

    block_on(async {
        let client = http_client_no_cookie_jar();
        let mut cookie = String::new();

        let register_resp = json_headers(
            client.post(format!("{}/api/auth/register", base_url)),
            &base_url,
            None,
        )
        .header("Content-Type", "application/json")
        .body(
            json!({
                "username": username,
                "password": password,
                "turnstileToken": "",
            })
            .to_string(),
        )
        .send()
        .await
        .map_err(|e| format!("zhujump register: {}", e))?;
        if !register_resp.status().is_success() {
            return Err(format!("zhujump register: {}", register_resp.status()));
        }
        cookie = merge_set_cookies(&cookie, register_resp.headers());
        let _ = register_resp.text().await;

        let csrf_resp = json_headers(
            client.get(format!("{}/api/auth/csrf", base_url)),
            &base_url,
            Some(&cookie),
        )
        .send()
        .await
        .map_err(|e| format!("zhujump csrf: {}", e))?;
        if !csrf_resp.status().is_success() {
            return Err(format!("zhujump csrf: {}", csrf_resp.status()));
        }
        cookie = merge_set_cookies(&cookie, csrf_resp.headers());
        let csrf_json: Value = csrf_resp
            .json()
            .await
            .map_err(|e| format!("zhujump csrf: {}", e))?;
        let csrf = csrf_json
            .get("csrfToken")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_string();
        if csrf.is_empty() {
            return Err("zhujump: csrf token missing".into());
        }

        let login_body = format!(
            "username={}&password={}&turnstileToken=&redirect=false&csrfToken={}&callbackUrl={}",
            urlencoding::encode(&username),
            urlencoding::encode(&password),
            urlencoding::encode(&csrf),
            urlencoding::encode(&login_referer(&base_url)),
        );
        let login_resp = json_headers(
            client.post(format!("{}/api/auth/callback/credentials?", base_url)),
            &base_url,
            Some(&cookie),
        )
        .header("Content-Type", "application/x-www-form-urlencoded")
        .header("x-auth-return-redirect", "1")
        .body(login_body)
        .send()
        .await
        .map_err(|e| format!("zhujump login: {}", e))?;
        if !login_resp.status().is_success() {
            return Err(format!("zhujump login: {}", login_resp.status()));
        }
        cookie = merge_set_cookies(&cookie, login_resp.headers());
        let _ = login_resp.text().await;

        let session_resp = json_headers(
            client.get(format!("{}/api/auth/session", base_url)),
            &base_url,
            Some(&cookie),
        )
        .send()
        .await
        .map_err(|e| format!("zhujump session: {}", e))?;
        if !session_resp.status().is_success() {
            return Err(format!("zhujump session: {}", session_resp.status()));
        }
        let session_json: Value = session_resp
            .json()
            .await
            .map_err(|e| format!("zhujump session: {}", e))?;
        let current_username = session_json
            .get("user")
            .and_then(|v| v.get("username"))
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_string();
        if current_username != username {
            return Err("zhujump: login verification failed".into());
        }

        let mut generate_body = json!({
            "name": local,
            "domain": domain,
        });
        if let Value::Object(ref mut obj) = generate_body {
            if let Some(expiry_time) = expiry_time {
                obj.insert("expiryTime".to_string(), Value::Number(expiry_time.into()));
            }
        }

        let generate_resp = json_headers(
            client.post(format!("{}/api/emails/generate", base_url)),
            &base_url,
            Some(&cookie),
        )
        .header("Content-Type", "application/json")
        .body(generate_body.to_string())
        .send()
        .await
        .map_err(|e| format!("zhujump generate: {}", e))?;
        if !generate_resp.status().is_success() {
            return Err(format!("zhujump generate: {}", generate_resp.status()));
        }
        let generate_json: Value = generate_resp
            .json()
            .await
            .map_err(|e| format!("zhujump generate: {}", e))?;
        let email = generate_json
            .get("email")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_string();
        let email_id = generate_json
            .get("id")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_string();
        if email.is_empty() || email_id.is_empty() {
            return Err("zhujump: invalid generate response".into());
        }
        let token = encode_token(&cookie, &email_id, &base_url)?;
        Ok(EmailInfo {
            channel,
            email,
            token: Some(token),
            expires_at: None,
            created_at: None,
        })
    })
}

pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let session = decode_token(token)?;
    let recipient = email.trim().to_string();
    if recipient.is_empty() {
        return Err("zhujump: empty email".into());
    }

    block_on(async {
        let base_url = session
            .b
            .as_deref()
            .unwrap_or(BASE_URL)
            .trim_end_matches('/')
            .to_string();
        let resp = json_headers(
            http_client_no_cookie_jar().get(format!(
                "{}/api/emails/{}",
                base_url,
                urlencoding::encode(&session.i)
            )),
            &base_url,
            Some(&session.c),
        )
        .send()
        .await
        .map_err(|e| format!("zhujump get emails: {}", e))?;
        if !resp.status().is_success() {
            return Err(format!("zhujump get emails: {}", resp.status()));
        }
        let data: Value = resp
            .json()
            .await
            .map_err(|e| format!("zhujump get emails: {}", e))?;
        let rows = data
            .get("messages")
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();

        let mut out = Vec::new();
        for row in rows {
            let mut obj = match row {
                Value::Object(map) => map,
                _ => continue,
            };
            let message_id = obj
                .get("id")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .trim()
                .to_string();
            let has_body = !obj
                .get("content")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .trim()
                .is_empty()
                || !obj
                    .get("html")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .trim()
                    .is_empty();
            if !message_id.is_empty() && !has_body {
                let detail_resp = json_headers(
                    http_client_no_cookie_jar().get(format!(
                        "{}/api/emails/{}/{}",
                        base_url,
                        urlencoding::encode(&session.i),
                        urlencoding::encode(&message_id)
                    )),
                    &base_url,
                    Some(&session.c),
                )
                .send()
                .await;
                if let Ok(detail_resp) = detail_resp {
                    if detail_resp.status().is_success() {
                        if let Ok(Value::Object(mut detail_json)) =
                            detail_resp.json::<Value>().await
                        {
                            if let Some(Value::Object(detail)) = detail_json.remove("message") {
                                for (k, v) in detail {
                                    obj.insert(k, v);
                                }
                            }
                        }
                    }
                }
            }
            let missing_to = obj
                .get("to_address")
                .map(|v| v.is_null() || v.as_str().unwrap_or("").trim().is_empty())
                .unwrap_or(true);
            if missing_to {
                obj.insert("to_address".to_string(), Value::String(recipient.clone()));
            }
            out.push(normalize_email(&Value::Object(obj), &recipient));
        }
        Ok(out)
    })
}
